core/helper_api.php

```diff
@@ -725,3 +725,35 @@ function helper_url_combine( $p_page, $p_query_string ) {
 
 	return $t_url;
 }
+
+/**
+ * Generate a hash to be used with dynamically generated content that is expected
+ * to be cached by the browser. This hash can be used to differentiate the generated
+ * content when it may be different based on some runtime attributes like: current user,
+ * project or language.
+ * An optional custom string can be provided to be added to the hash, for additional
+ * differentiating criteria, but this string must be already prepared by the caller.
+ *
+ * @param array $p_runtime_attrs    Array of attributes to be calculated from current session.
+ *                                  possible values: 'user', 'project', 'lang'
+ * @param string $p_custom_string   Additional string provided by the caller
+ * @return string                   A hashed md5 string
+ */
+function helper_generate_cache_key( array $p_runtime_attrs = [], $p_custom_string = '' ) {
+	# always add core version, to force reload of resources after an upgrade.
+	$t_key = $p_custom_string . '+V' . MANTIS_VERSION;
+	$t_user_auth = auth_is_user_authenticated();
+	foreach( $p_runtime_attrs as $t_attr ) {
+		switch( $t_attr ) {
+			case 'user':
+				$t_key .= '+U' . ( $t_user_auth ? auth_get_current_user_id() : META_FILTER_NONE );
+				break;
+			case 'project':
+				$t_key .= '+P' . ( $t_user_auth ? helper_get_current_project() : META_FILTER_NONE );
+				break;
+			case 'lang':
+				$t_key .= '+L' . lang_get_current();
+		}
```

Comment:
Add a default that triggers a generic error.

Reply:
> Add a default that triggers a generic error.

added